File: q101-125/q101.py
# 2020.01.09
# Problem Statement:
# https://leetcode.com/problems/symmetric-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        # left: pre-order traversal
        # right: post-order traversal
        # just check if left and right are "same" with different traverse orders
        # can use the same function as question 100
        
        if not root: return True
        if root.left is None: return root.right is None
        if root.right is None: return root.left is None
        
        left_root, right_root = root.left, root.right
        self.answer = True
        return self.isSymmetricHelper(left_root, right_root)
        
    def isSymmetricHelper(self, p, q):
        # one of the tree is already went through
        if p is None: 
            self.answer = self.answer and (q is None)
            return self.answer
        if q is None:
            self.answer = self.answer and (p is None)
            return self.answer
        
        # p in order traversal, q post order traversal
        # please note the opposition of left and right
        if p.left is not None: 
            self.isSymmetricHelper(p.left, q.right)
        else:
            # check if q has extra left subtree
            self.answer = self.answer and (q.right is None)
        
        # do the value comparison
        self.answer = self.answer and (p.val == q.val)
        
        if p.right is not None:
            self.isSymmetricHelper(p.right, q.left)
        else:
            # check if q has extra right subtree
            self.answer = self.answer and (q.left is None)
        
        return self.answer
        
        

File: Amazon/q101.java
// 2022.05.27
// Problem Statement:
// https://leetcode.com/problems/symmetric-tree/

// idea: recursion, use a helper function to check if 2 subtrees are symmetric,
// left subtree's right subtree should be symmetric to right subtree's left subtree,
// left subtree's left subtree should be symmetric to right subtree's right subtree,
// otherwise not a symmetric tree
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetricHelper(TreeNode root1, TreeNode root2) {
        if (root1==null && root2==null) return true;
        if (root1==null || root2==null) return false;
        if (root1.val!=root2.val) {
            return false;
        }
        boolean ret = isSymmetricHelper(root1.left, root2.right);
        if (!ret) return false;
        ret = ret && isSymmetricHelper(root1.right, root2.left);
        return ret;
    }
    
    public boolean isSymmetric(TreeNode root) {
        if (root==null) return true;
        return isSymmetricHelper(root.left, root.right);
    }
}